### PR TITLE
Stabilize `ExternalDocumentableProvider`

### PR DESCRIPTION
**Pull request by IgnatBeresnev**
Part of #3112, depends on #3195

Creating this PR as a draft because I still need to write documentation for the new API, but the review can already start as I don't plan to introduce any significant code changes.

### CODE REVIEW

File: subprojects/analysis-kotlin-api/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/documentable/ExternalDocumentableProvider.kt [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+package org.jetbrains.dokka.analysis.kotlin.documentable
+
+import org.jetbrains.dokka.DokkaConfiguration.DokkaSourceSet
+import org.jetbrains.dokka.links.DRI
+import org.jetbrains.dokka.model.DClasslike
+
+/**
+ * TODO [external-documentable-provider] add documentation
+ */
+public interface ExternalDocumentableProvider {
```
Comment (IgnatBeresnev) [inline subprojects/analysis-kotlin-api/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/documentable/ExternalDocumentableProvider.kt:14]: In a few words, this extension is used in dackka to create documentables for supertypes that come from dependencies, because, if I'm not mistaken, by default Dokka only creates the documentable model for the code it finds in the project itself, but not its classpath.

File: subprojects/analysis-kotlin-api/src/test/kotlin/org/jetbrains/dokka/analysis/test/ExternalDocumentableProviderTest.kt [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+
+import org.jetbrains.dokka.analysis.test.api.javaTestProject
+import org.jetbrains.dokka.analysis.test.api.kotlinJvmTestProject
+import org.jetbrains.dokka.analysis.test.api.useServices
+import org.jetbrains.dokka.analysis.test.api.util.getResourceAbsolutePath
+import org.jetbrains.dokka.links.DRI
+import org.jetbrains.dokka.model.*
+import kotlin.test.*
+
+class ExternalDocumentableProviderTest {
```
Comment (vmishenev) [inline subprojects/analysis-kotlin-api/src/test/kotlin/org/jetbrains/dokka/analysis/test/ExternalDocumentableProviderTest.kt:15]: Is this yet another set of tests for external documentable? 
What are the plans for `ExternalDocumentablesTest`? It also contains interesting corner cases.

**Review by vmishenev — COMMENTED** [review]
> Creating this PR as a draft because I still need to write documentation for the new API, but the review can already start as I don't plan to introduce any significant code changes.

Ideal review is not only about code.

Is this just renaming `internal.ExternalDocumentablesProvider` into `ExternalDocumentableProvider`?

File: subprojects/analysis-kotlin-api/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/documentable/ExternalDocumentableProvider.kt (hunk 1 above)
Comment (whyoleg) [reply subprojects/analysis-kotlin-api/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/documentable/ExternalDocumentableProvider.kt:14]: BTW, what about typealiases? They are not `DClasslike` but could be supertypes, and so do we need to support them?

Comment (IgnatBeresnev) [reply subprojects/analysis-kotlin-api/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/documentable/ExternalDocumentableProvider.kt]: It could indeed be useful, but we'd have to think about the corner cases, write tests for it and make sure the implementations support it, and at this point it doesn't seem to be needed, at least not for #3112. If someone does come with such a request, we can certainly consider it

File: subprojects/analysis-kotlin-api/src/test/kotlin/org/jetbrains/dokka/analysis/test/ExternalDocumentableProviderTest.kt (hunk 2 above)
Comment (IgnatBeresnev) [reply subprojects/analysis-kotlin-api/src/test/kotlin/org/jetbrains/dokka/analysis/test/ExternalDocumentableProviderTest.kt]: Deleted the old tests, as the new tests should cover all sensible corner cases I could think of

**Comment by IgnatBeresnev** [comment]
> Is this just renaming internal.ExternalDocumentablesProvider into ExternalDocumentableProvider?

Not only that, but adding comprehensible documentation and unit tests as well. The implementation didn't change though.

**Comment by IgnatBeresnev** [comment]
Had to rebase the branch onto master and resolve some merge conflicts manually, so I squashed the commits and force pushed it.

It's ready for review now

**Review by whyoleg — APPROVED** [review]

**Review by vmishenev — APPROVED** [review]

File: subprojects/analysis-kotlin-api/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/documentable/ExternalDocumentableProvider.kt (hunk 1 above)
Comment (vmishenev) [reply subprojects/analysis-kotlin-api/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/documentable/ExternalDocumentableProvider.kt]: > BTW, what about typealiases? They are not DClasslike but could be supertypes, and so do we need to support them?

AFAIK, [dokka-devsite-plugin](https://android.googlesource.com/platform/tools/dokka-devsite-plugin) can use `TypeAliased` that has `inner`.
